### PR TITLE
Leaves AND Branches - Elizabeth AND Macaria

### PR DESCRIPTION
**Pull request by north108**
# Video Store API
Congratulations! You're submitting your assignment!
If you didn't get to the functionality the question is asking about, reply with what you would have done if you had completed it.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Explain how you came up with the design of your ERD, based on the seed data. | The seed data informed us on which columns to include in our schema and made us think about how we were going to connect movies to customers and customers to movies. 
What would be the Big-O time complexity of your /customers & /movies endpoints? What does the time complexity depend on? Explain your reasoning. | O(n) n being the number of customers and movies because it has to loop through each element and return the attributes of each element.
What is the Big-O time complexity of the `POST /rentals/check-in` endpoint? What does the time complexity depend on? Explain your reasoning.  | O(1) because we search using params which access hashes. 
Describe a set of positive and negative test cases you implemented for a model. | In the movie model tests, we tested that the movie is valid with the  presence of a title and invalid without a title.
Describe a set of positive and negative test cases you implemented for a controller. | In the movie controller test #show we tested that it returned the movie with a valid id and responds with a bad request if given an invalid id.
How does your API respond when bad data is sent to it? | Bad request error.
Describe one of your custom model methods and why you chose to wrap that functionality into a method. | In the rental model we wrote a method called due_date and put it in a model method because it is business logic.
Do you have any recommendations on how we could improve this project for the next cohort? | Smoke tests led us to have write non-optimal code.
Link to Trello | We pair programmed the whole time so we didn't have the need for one.
Link to ERD | https://imgur.com/a/IVHIglv
Summary | Elizabeth was pleasure to work with! A true dynamo! Her calm attitude and efficient work style made this project a dream for me! Also, I liked the size of the project for a pair. We were able to do most of the work together without taking too much work home throughout the week. True pair programming! 
Macaria was a dream, as per usual!

### CODE REVIEW

File: app/controllers/rentals_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+      movie = Movie.find_by(id: rental_params[:movie_id])
+
+      if (!movie.available_inventory)
+        movie.available_inventory = movie.inventory
+      end
+
+      movie.available_inventory -= 1
+      movie.available_inventory_will_change!
+      movie.save
```
Comment (CheezItMan) [inline app/controllers/rentals_controller.rb:19]: This kind of business logic looks like a prime candidate for a model method!

File: app/controllers/rentals_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      if (!movie.available_inventory)
+        movie.available_inventory = movie.inventory
+      end
+
+      movie.available_inventory += 1
+      movie.available_inventory_will_change!
+      movie.save
+      
+      rental.status = "checked_in"
+      rental.save
```
Comment (CheezItMan) [inline app/controllers/rentals_controller.rb:46]: This is more business logic to put in the model.

File: app/models/customer.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  validates :name, presence: true
+
+  def self.check_out_movie(id)
+    customer = Customer.find_by(id: id)
+    movie_count = customer.movies_checked_out_count
+    customer.update(movies_checked_out_count: movie_count + 1)
+
+    customer.save
+    return customer
+  end
```
Comment (CheezItMan) [inline app/models/customer.rb:13]: So the `checked_out_count` has nothing to do with the number of `Rentals` the customer has.  Hmmm... Any kind of subtle bug could get these out of sync.  This isn't an out-and-out bug, but in an imperfect system it's an invitation for these kinds of problems.

File: app/models/rental.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+   rental.due_date = due_date
+   rental.save
+   return rental
+  end
+
+  def self.status_checkout(rental)
+    rental.status = "checked out"
+    rental.save
+    return rental
+  end
```
Comment (CheezItMan) [inline app/models/rental.rb:18]: These make much more sense as instance methods as opposed to class methods.

File: app/models/movie.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class Movie < ApplicationRecord
+  has_many :rentals, dependent: :nullify
+  validates :title,  presence: true
+  validates :inventory, presence: true, numericality: {greater_than: 0}
```
Comment (CheezItMan) [inline app/models/movie.rb:4]: Running the seeds file, I'm getting validation errors, so your seeds file doesn't work with this validation.

File: app/controllers/movies_controller.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+  def create
+    new_movie = Movie.new(movie_params)
+    new_movie.available_inventory = movie_params[:inventory]
+
+    if new_movie.save
+      render json: new_movie.as_json(only: [:id]), status: :ok
+      return
+    else
+      render json: { errors: ["BAD Request"] }, status: :bad_request
```
Comment (CheezItMan) [inline app/controllers/movies_controller.rb:31]: So you're not reporting validation errors?

File: test/models/customer_test.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
+require "test_helper"
+
+describe Customer do
+  describe "relations" do
+    let(:current_customer) { customers(:customer1)}
+
+    it "can get the list of rentals from 'rentals'" do
```
Comment (CheezItMan) [inline test/models/customer_test.rb:7]: You should have negative and positive cases of each relationship (with rentals and without).

File: test/models/customer_test.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      customer = customers(:customer1)
+
+      result = customer.valid?
+
+      expect(result).must_equal true
+    end
+  end
+
+  describe "check_out_movie" do 
+    it "increases the customers movie count if given valid data" do
```
Comment (CheezItMan) [inline test/models/customer_test.rb:35]: You  should have a test with invalid data.  Also this method doesn't make much sense in that someone could have a larger checked out count without more rentals.

File: test/models/rental_test.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    it "is valid with a numeric movie_id" do 
+      rental = rentals(:rental1)
+
+      result = rental.valid?
+
+      expect(result).must_equal true
+    end
+
+    it "is invalid with a movie_id that is not a number" do
```
Comment (CheezItMan) [inline test/models/rental_test.rb:74]: You should also have a test so that it's invalid to check out a movie, if there are none left in inventory.

File: test/controllers/movies_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
+require "test_helper"
+
+MOVIE_FIELDS = ["available_inventory", "inventory", "overview", "release_date", "title"]
+
+describe MoviesController do
+  describe "index" do
```
Comment (CheezItMan) [inline test/controllers/movies_controller_test.rb:6]: 👍

File: test/controllers/movies_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+      no_title_movie = @new_movie
+      no_title_movie[:title] = nil
+
+      expect { 
+        post movies_path(@new_movie)
+      }.wont_change "Movie.count"
+
+      must_respond_with :bad_request
+      body = JSON.parse(response.body)
+      expect(body['errors']).must_include "BAD Request"
```
Comment (CheezItMan) [inline test/controllers/movies_controller_test.rb:89]: You should also have some validation error messages.

File: test/controllers/rentals_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        post check_out_path(new_rental)
+      }.wont_change "Rental.count"
+
+      must_respond_with :bad_request
+      body = JSON.parse(response.body)
+      expect(body['errors']).must_include "Bad Request"
+    end
+  end
+
+  describe "check_in" do
```
Comment (CheezItMan) [inline test/controllers/rentals_controller_test.rb:37]: What about invalid check ins and checkouts?  Trying to check out a movie with no more inventory, or checking in a movie that's not checked out.  More testing needed.

**Comment by CheezItMan** [comment]
# Video Store
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene |  Good number of commits and mostly granular commits
Comprehension questions |  For the check-in endpoint you need to look up the proper movie and customer, so... this will depend on how this data is organized O(n) worst-case if the database has to do a linear search, but more likely O(1) since the database will use something like a hash to index the values. 
**General** |
Business Logic in Models |  Overall pretty good
All required endpoints return expected JSON data |  Check
Requests respond with appropriate HTTP Status Code | Check
Errors are reported |  Error status is reported, but no validation errors reported.
**Testing** |
Passes all Smoke Tests | Check
Model Tests - all relations, validations, and custom functions test positive & negative cases | See my inline comments, overall ok
Controller Tests - URI parameters and data in the request body have positive & negative cases | You're not covering all your edge cases here. 
**Overall** |  It works, but it's quite a bit unpolished.  Check out my inline comments and let me know if you have questions.
